src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxValueProvider_ForAttributeWithSimpleName.cs

```diff
@@ -83,19 +87,23 @@ public partial struct SyntaxValueProvider
             .Select((tuple, _) => GlobalAliases.Concat(tuple.Left, tuple.Right))
             .WithTrackingName("allUpIncludingCompilationGlobalAliases_ForAttribute");
 
-        // Combine the two providers so that we reanalyze every file if the global aliases change, or we reanalyze a
-        // particular file when it's compilation unit changes.
-        var syntaxTreeAndGlobalAliasesProvider = syntaxTreesProvider
-            .Combine(allUpGlobalAliasesProvider)
-            .WithTrackingName("compilationUnitAndGlobalAliases_ForAttribute");
+        // Filter down to the trees that contain attributes in them.  in general, this will normally be smaller than the
+        // number of total trees, allowing us to hold onto a much smaller number of nodes.
+        var treesContainingAttribute = syntaxTreesProvider.Where(
+            (t, c) =>
+            {
+                // Walk the green node tree first to avoid allocating the entire red tree for files that have no attributes.
+                var root = t.GetRoot(c);
+                return ContainsAttributeList(root.Green, syntaxHelper.AttributeListKind);
+            });
 
-        // For each pair of compilation unit + global aliases, walk the compilation unit 
-        var result = syntaxTreeAndGlobalAliasesProvider
-            .SelectMany((globalAliasesAndCompilationUnit, cancellationToken) => GetMatchingNodes(
-                syntaxHelper, globalAliasesAndCompilationUnit.Right, globalAliasesAndCompilationUnit.Left, simpleName, predicate, cancellationToken))
-            .WithTrackingName("result_ForAttribute");
+        var rootsWithAttributeAndGlobalAliasesProvider = treesContainingAttribute.Combine(allUpGlobalAliasesProvider)
+            .WithTrackingName("compilationUnitAndGlobalAliases_ForAttribute");
 
-        return result;
+        return rootsWithAttributeAndGlobalAliasesProvider.Select(
+            (tuple, c) => GetMatchingNodes(syntaxHelper, tuple.Right, tuple.Left, simpleName, predicate, c))
+            .WithComparer(ImmutableArrayValueComparer<SyntaxNode>.Instance)
+            .WithTrackingName("result_ForAttributeInternal");
```

Comment:
note: importantly, no SelectMany here.  We're searching one syntax tree, and we want all the results from that single syntax tree to stay together.